### PR TITLE
gh-123080: Additional ast optimizations

### PR DESCRIPTION
**Pull request by wrongnull**
This allows some expression templates to produce less bytecode and perform more folding operations in the ```fold_compare ``` function.
For instance
```
>>> from dis import dis
>>> dis("1 in [1, 2, 3, 4]")
  0           RESUME                   0

  1           RETURN_CONST             0 (True)
>>> dis("1 == 1")
  0           RESUME                   0

  1           RETURN_CONST             0 (True)
>>> 
```

* Issue: gh-123080

### CODE REVIEW

**Comment by wrongnull** [comment]
I guess there is no need to update news for this

**Comment by terryjreedy** [comment]
I believe that changing bytecode requires News and What's New items.

**Comment by wrongnull** [comment]
I'll convert it to draft so far because of tests segfaults. Also I'll wait until https://github.com/python/cpython/pull/122667 will be merged

**Comment by wrongnull** [comment]
For now, everything works fine and all the previous optimizations are preserved. I'll wait for tests of ast optimizations to be merged.

File: Python/ast_opt.c [hunk 1]
```diff
@@ -624,20 +624,87 @@ static int
...
+        }
+
+        elts_len = asdl_seq_LEN(elts);
+        if (!elts_len) {
+            return make_const(node, op == In ? Py_False : Py_True, arena);
+        }
+
+        PyObject *newval = PyTuple_New(elts_len);
+        if (newval == NULL) {
+            if (PyErr_ExceptionMatches(PyExc_KeyboardInterrupt)) {
```
Comment (pablogsal) [inline Python/ast_opt.c]: There is no need to match against `PyExc_KeyboardInterrupt` here. Just treat this like a regular error and bubble up

File: Python/ast_opt.c [hunk 2]
```diff
@@ -624,20 +624,87 @@ static int
...
+                return 0;
+            }
+            PyErr_Clear();
+            return 1;
+        }
+
+        for (Py_ssize_t j = 0; j < elts_len; j++) {
+            expr_ty e = (expr_ty)asdl_seq_GET(elts, j);
+            if (e->kind != Constant_kind) {
+                PyObject_GC_UnTrack(newval);
```
Comment (pablogsal) [inline Python/ast_opt.c]: Why are you calling directly these GC functions? Just call a normal `Py_DECREF` when you are done with it

File: Python/ast_opt.c [hunk 3]
```diff
@@ -624,20 +624,87 @@ static int
...
+                PyObject_GC_Del(newval);
+                return 1;
+            }
+            PyObject *v = e->v.Constant.value;
+            PyTuple_SET_ITEM(newval, j, Py_NewRef(v));
+            if (is_lhs_constant &&
+                (v == (node->v.Compare.left->v.Constant.value)))
+            {
+                PyObject_GC_UnTrack(newval);
```
Comment (pablogsal) [inline Python/ast_opt.c]: Same as https://github.com/python/cpython/pull/123081/files#r1722232517

File: Python/ast_opt.c [hunk 4]
```diff
@@ -624,20 +624,87 @@ static int
...
+                PyObject_GC_Del(newval);
+                return make_const(node, op == In ? Py_True : Py_False, arena);
+            }
         }
+
+        if (arg->kind == Set_kind)
```
Comment (pablogsal) [inline Python/ast_opt.c]: PEP 7: Add braces

File: Python/ast_opt.c [hunk 5]
```diff
@@ -624,20 +624,87 @@ static int
...
+            Py_SETREF(newval, PyFrozenSet_New(newval));
```
Comment (pablogsal) [inline Python/ast_opt.c]: `PyFrozenSet_New` can fail, you need to check for errors and handle the failure

File: Python/ast_opt.c [hunk 6]
```diff
@@ -624,20 +624,87 @@ static int
...
     int op = asdl_seq_GET(ops, i);
+    arg = asdl_seq_GET(args, i);
+    _Bool is_lhs_constant = (node->v.Compare.left->kind) == Constant_kind;
+    //lhs_const = node->v.Compare.left->v.Constant.value;
     if (op == In || op == NotIn) {
-        if (!fold_iter((expr_ty)asdl_seq_GET(args, i), arena, state)) {
-            return 0;
+        if (arg->kind == List_kind) {
+            asdl_expr_seq *list_elts = arg->v.List.elts;
+            if (has_starred(list_elts))
```
Comment (pablogsal) [inline Python/ast_opt.c:644]: PEP 7: add braces

File: Python/ast_opt.c [hunk 7]
```diff
@@ -624,20 +624,87 @@ static int
...
+
+        if (arg->kind == Set_kind)
+            Py_SETREF(newval, PyFrozenSet_New(newval));
+        return make_const(arg, newval, arena);
     }
+    else if ((op == Eq || op == NotEq) &&
+             (arg->kind == Constant_kind) &&
+             is_lhs_constant)
+    {
+        if ((node->v.Compare.left->v.Constant.value) == (arg->v.Constant.value))
```
Comment (pablogsal) [inline Python/ast_opt.c]: PEP 7: Add braces

File: Python/ast_opt.c [hunk 8]
```diff
@@ -624,20 +624,87 @@ static int
...
 
     ops = node->v.Compare.ops;
     args = node->v.Compare.comparators;
     /* Change literal list or set in 'in' or 'not in' into
        tuple or frozenset respectively. */
     i = asdl_seq_LEN(ops) - 1;
     int op = asdl_seq_GET(ops, i);
+    arg = asdl_seq_GET(args, i);
+    _Bool is_lhs_constant = (node->v.Compare.left->kind) == Constant_kind;
+    //lhs_const = node->v.Compare.left->v.Constant.value;
```
Comment (pablogsal) [inline Python/ast_opt.c]: Why is this commented out?

**Review by pablogsal — CHANGES_REQUESTED** [review]
Left some comments

File: Python/ast_opt.c (hunk 8 above)
Comment (wrongnull) [reply Python/ast_opt.c]: Sorry I leave it by accident

File: Python/ast_opt.c [hunk 9]
```diff
@@ -624,20 +624,87 @@ static int
...
+            }
+            PyErr_Clear();
+            return 1;
+        }
+
+        for (Py_ssize_t j = 0; j < elts_len; j++) {
+            expr_ty e = (expr_ty)asdl_seq_GET(elts, j);
+            if (e->kind != Constant_kind) {
+                PyObject_GC_UnTrack(newval);
+                PyObject_GC_Del(newval);
```
Comment (Eclips4) [inline Python/ast_opt.c]:
```suggestion
                Py_DECREF(newval);
```

File: Python/ast_opt.c (hunk 5 above)
Comment (Eclips4) [reply Python/ast_opt.c]: This line was copied from `fold_iter`. So, `fold_iter` has the same issue (no error handling for NULL values on calls to `PyFrozenSet_New`). Do you think it's OK to address this issue in this PR?

**Comment by pablogsal** [comment]
> Do you think it's OK to address this issue in this PR?

I prefer to do that separately because we want to back port it. I will take care of that

**Pull request closed without merging**